Algolia migration guide (#1623)
meilisearch/documentation: learn/getting_started/algolia_migration.md

# Migrating from Algolia to Meilisearch

This guide will walk you through exporting your Algolia data and indexing it in Meilisearch using a Node.js script. To skip directly to the finished script, [click here](#finished-script).

This guide also includes comparisons of Meilisearch and Algolia's [APIs](#apis), [index settings, parameters](#settings-and-parameters), and [support for the `instantsearch.js` library](#front-end-components).

This is a practical guide for current Algolia users making the transition to Meilisearch. We also provide [a high-level comparison of the two search companies and their products](/learn/what_is_meilisearch/comparison_to_alternatives.md#meilisearch-vs-algolia) elsewhere.

## Migration script

Migrating indexes from Algolia to Meilisearch is a straightforward process:

> 1. Export your data stored in Algolia

Reviewer: Can we use each of these steps as an H2 and use the H3s as subheadings for this?

Author: Great idea!

2. Upload your data to Meilisearch
3. (optional) Configure your settings to approximate the settings of your Algolia index

> This guide uses a [Node.js](https://nodejs.org/en/) script to upload Algolia index data to Meilisearch. If you would like to write this script in another language, we provide [SDKs for most common languages](/learn/what_is_meilisearch/sdks.md) and [welcome contributions to this guide](https://github.com/meilisearch/documentation/edit/master/learn/getting_started/algolia_migration.md).

Reviewer: ```suggestion
This guide uses a [Node.js](https://nodejs.org/en/) script to upload Algolia index data to Meilisearch. If you would like to write this script in another language, we provide [SDKs](/learn/what_is_meilisearch/sdks.md) for most common languages and welcome [contributions to this guide](https://github.com/meilisearch/documentation/edit/master/learn/getting_started/algolia_migration.md).
```

> Before continuing, make sure you have both Meilisearch and Node.js installed, and access to a command-line terminal. If you're unsure, see our [quick start](/learn/getting_started/quick_start.md) for instructions on installing Meilisearch.

Reviewer: ```suggestion
Before continuing, make sure you have both Meilisearch and Node.js installed and have access to a command-line terminal. If you're unsure, see our [quick start](/learn/getting_started/quick_start.MD) for instructions on installing Meilisearch.
```

> Before continuing, make sure you have both Meilisearch and Node.js installed, and access to a command-line terminal. If you're unsure, see our [quick start](/learn/getting_started/quick_start.md) for instructions on installing Meilisearch.

Reviewer 2: To be safe, we can mention the current node version as well as the versions of meilisearch and algoliasearch dependencies that we are using.

Node v16.14
algoliasearch 4.13
meilisearch  0.25

> Before continuing, make sure you have both Meilisearch and Node.js installed, and access to a command-line terminal. If you're unsure, see our [quick start](/learn/getting_started/quick_start.md) for instructions on installing Meilisearch.

Reviewer 3: It is always best to mention the exact versions to help the user debug and also to avoid having obsolete guides. With the specified versions we guarantee to the user that this guide works in these circumstances. 

I think we should also add the version of meilisearch (the engine)

### Initialize project

Start by creating a directory `algolia-meilisearch-migration` and initializing it as an npm package:

```bash
mkdir algolia-meilisearch-migration
cd algolia-meilisearch-migration
npm init -y
```

This will set up the environment we need to install dependencies.

Next, create a `script.js` file:

```bash
touch script.js
```

This file will contain our migration script.

### Install required packages

To get started, you'll need two different packages. The first is `algoliasearch`, the JavaScript client for the Algolia API, and the second is `meilisearch`, the JavaScript client for the Meilisearch API.

```bash

> npm install algoliasearch meilisearch

Reviewer 4: It's been a while, but I believe it's good practice to add a `-s` flag to save these dependencies to `packages.json`. What do you think, @bidoubiwa?

```

### Create Algolia client

> You'll need your **Application ID** and **API Key** to start the Algolia client. Both can be found in your [Algolia account](https://www.algolia.com/api-keys).

Reviewer: Should this be API key or admin API key?

Author: Not sure what you mean, could you clarify?

Reviewer: This line says the user needs an application ID and an API key. Line 68 uses the term "admin API key". If they're talking about the same thing, we should use the same terms (whichever is correct)

Author: Nice eye! Thanks for the clarification 😊

Paste the below code in `script.js`:

```js

> const AlgoliaSearch = require("algoliasearch");

Reviewer 4: In JS, capitalised names indicate functions that should be called with the `new` keyword. Though this is only a convention, I think we should not break it without a good reason.

```suggestion
const algoliaSearch = require("algoliasearch");
```

const algoliaClient = AlgoliaSearch(
  "APPLICATION_ID",
  "ADMIN_API_KEY"
);
const algoliaIndex = algoliaClient.initIndex("INDEX_NAME");
```

Replace `APPLICATION_ID` and `ADMIN_API_KEY` with your Algolia application ID and admin API key respectively.

Replace `INDEX_NAME` with the name of the Algolia index you would like to migrate to Meilisearch.

### Fetch data from Algolia

To fetch all Algolia index data at once, use Algolia's [`browseObjects`](https://www.algolia.com/doc/api-reference/api-methods/browse/) method.

```js
let records = [];
await algoliaIndex.browseObjects({
    batch: (hits) => {

>       records=records.concat(hits)

Reviewer 4: ```suggestion
      records = records.concat(hits);
```

    }
  });
```

> On each batch of hits, the `batch` callback method is invoked, and the content is concatenated in the `records` array. This variable will be used later on in the upload process.

Reviewer: ```suggestion
The `batch` callback method is invoked on each batch of hits, and the content is concatenated in the `records` array. This variable will be used later on in the upload process.
```

> On each batch of hits, the `batch` callback method is invoked, and the content is concatenated in the `records` array. This variable will be used later on in the upload process.

Reviewer: Do you mean `records` when you say "this variable"? I think it would be better to specify it

> On each batch of hits, the `batch` callback method is invoked, and the content is concatenated in the `records` array. This variable will be used later on in the upload process.

Reviewer 4: Nitpick: a full paragraph using only the passive voice 😅

> #### Nested objects

Reviewer: Maybe we can split this section into "v0.27 or above" and "v0.26 and below". That way, the user knows what part is for them instead of reading the whole section

> Meilisearch versions prior to v0.27.0 have limited support for nested fields, meaning that **nested objects must be flattened manually**. This can be avoided entirely by using [the latest version of Meilisearch](https://github.com/meilisearch/MeiliSearch/releases). If this is impossible for some reason, try the modified data retrieval process below.

Reviewer 4: I prefer to leave patch versioning outside of versions unless the patch is relevant to the subject.

```suggestion
Meilisearch versions prior to v0.27 have limited support for nested fields, meaning that **nested objects must be flattened manually**. This can be avoided entirely by using [the latest version of Meilisearch](https://github.com/meilisearch/MeiliSearch/releases). If this is impossible for some reason, try the modified data retrieval process below.
```

Let's say your document structure looks like this:

```json
{
    "objectID": 1,
    "name": "John",
    "age": 20,
    "address": {
        "country": "Spain"
    }
}
```

To export this data from Algolia and simultaneously flatten it, use the same `browseObjects` method as described above. However, before concatenation, flatten the `address` object into `address.country`, and perform the same operation for other subfields:

```js
let records = [];
await algoliaIndex.browseObjects({
 batch: (recordsArr) => {
        records = recordsArr.reduce((acc, record) => {
          const { address, ...rest } = record;
          acc.push({ "address.country": address.country, ...rest });
          return acc;
        }, records);
      }
});
```

> Adapt this principle to your own dataset by replacing `address` and `address.country` with a parent object and nested field, respectively.
> 

Reviewer 4: I think we could cut this paragraph.

```suggestion
```

### Create Meilisearch client

> Create a Meilisearch client by passing the host URL and API key of your Meilisearch instance. The easiest option is to use the automatically generated [admin API key](/learn/security/master_api_keys.md#listing-api-keys).

Reviewer 4: I would appreciate a brief sentence reminding me why I need to create a Meilisearch client.

I know, it's sort of obvious, but I think it helps the user knowing where they are in the process.

```js
const { MeiliSearch } = require("meilisearch");
const meiliClient = new MeiliSearch({
  host: "MEILI_HOST",
  apiKey: "MEILI_API_KEY",

> });

Reviewer 4: Just for reference, I mentioned capitalisation as a JS convention to indicate a function should be invoked with `new` and how `const AlgoliaSearch` broke that convention. 

Here you see that the constant `Meilisearch` is capitalised and then invoked as `new Meilisearch ([etc])`, which is what I would expect as a JS programmer.

const meiliIndex = meiliClient.index("MEILI_INDEX_NAME");
```

> Replace `MEILI_HOST`,`MEILI_API_KEY`, and `MEILI_INDEX_NAME` with your Meilisearch host URL, Meilisearch API key, and the index name where you would like documents to be added. Meilisearch will create the index if it doesn't already exist.

Reviewer 4: ```suggestion
Replace `MEILI_HOST`,`MEILI_API_KEY`, and `MEILI_INDEX_NAME` with your Meilisearch host URL, Meilisearch API key, and the index name where you would like to add documents. Meilisearch will create the index if it doesn't already exist.
```

### Upload data to Meilisearch

> Next, use the Meilisearch [`addDocumentsInBatches`](/learn/core_concepts/documents.md#primary-field) method to upload all your records in batches of 1000 at a time.

Reviewer 4: Is this "at a time" redundant?

```suggestion
Next, use the Meilisearch [`addDocumentsInBatches`](/learn/core_concepts/documents.md#primary-field) method to upload all your records in batches of 1000.
```

| filters | filter |
| facets | facetsDistribution |
| attributesToHighlight | attributesToHighlight |
| offset | offset |
| length | limit |
| typoTolerance | Coming in Meilisearch v0.27. It will also support Algolia’s  minWordSizefor1Typo and minWordSizefor2Typos options. |
| snippetEllipsisText | Coming in Meilisearch v0.27 |
| searchableAttributes | searchableAttributes |
| attributesForFaceting | filterableAttributes |
| unretrievableAttributes | No direct equivalent; is achieved by removing attributes from displayedAttributes. |
| attributesToRetrieve | displayedAttributes |
| attributeForDistinct | distinctAttribute |
| ranking | rankingRules |
| Custom Ranking | Custom Ranking |
| removeStopWords | stopWords |
| synonyms | synonyms |
| Sorting(using replicas) | sortableAttributes(no replicas required)  |
| removeWordsIfNoResults | Meilisearch automatically supports it. However, it isn't customizable. |
| disableTypoToleranceOnAttributes | Coming in Meilisearch v0.27 |
| separatorsToIndex | Not Supported |
| disablePrefixOnAttributes | Not Supported |
| relevancyStrictness | Not Supported |
| maxValuesPerFacet | Not Supported |
| sortFacetValuesBy | Not Supported |

> | restrictHighlightAndSnippetArrays | Not Supported |

Reviewer 4: We really need a markdown table auto-formatter.

## APIs

This section compares Algolia and Meilisearch's respective API methods, using JavaScript for reference.

> | Topic | Algolia | Meilisearch |

Reviewer: Not sure if "Topic" is the correct term here

| Add Object | `index.saveObjects(array objects)` | `index.addDocuments(array objects)` |
| Partial Update Object | `index.partialUpdateObjects(array objects)` | `index.updateDocuments(array objects)` |
| Delete All Objects | `index.deleteObjects(array objectIDs)` | `index.deleteAllDocuments()` |
| Delete One Object | `index.deleteObject(string objectID)` | `index.deleteDocument(string id)` |
| Get All Objects | `index.getObjects(array objectIDs)` | `index.getDocuments(object params)` |
| Get Single Object | `index.getObject(str objectID)` | `index.getDocument(string id)` |
| Get API Keys | `client.listApiKeys()` | `client.getKeys()` |
| Get API Key Info | `client.getApiKey(string apiKey)` | `client.getKey(string apiKey)` |
| Create API Key | `client.addApiKey(array acl)` | `client.createKey(object configuration)` |
| Update API Key | `client.updateApiKey(string apiKey, object configuration)` | `client.updateKey(string apiKey, object configuration)` |
| Delete API Key | `client.deleteApiKey(string apiKey)` | `client.deleteKey(string apiKey)` |

> ## Front-end components

Reviewer: This section feels incomplete, maybe use a table here instead?

Author: Do you mean front-end components? There used to be a table here, but we removed it after deciding it would be too muchtrouble to keep up-to-date. decided it would be too much trouble to keep it up-to-date. Probably better to send people to Instant Meilisearch's repo/documentation anyway.

[InstantSearch](https://github.com/algolia/instantsearch.js) is a collection of open-source tools maintained by Algolia and used to generate front-end search UI components.

> [Instant Meilisearch](https://github.com/meilisearch/instant-meilisearch) is a plugin connecting your Meilisearch instance with InstantSearch, giving you access to many (but not all) of the same front-end components as Algolia users. Here is [an up-to-date list of components compatible with Instant Meilisearch](https://github.com/meilisearch/instant-meilisearch/#-api-resources).

Reviewer: ```suggestion
[Instant Meilisearch](https://github.com/meilisearch/instant-meilisearch) is a plugin connecting your Meilisearch instance with InstantSearch, giving you access to many (but not all) of the same front-end components as Algolia users. [Here](https://github.com/meilisearch/instant-meilisearch/#-api-resources) is an up-to-date list of components compatible with Instant Meilisearch.
```

> [Instant Meilisearch](https://github.com/meilisearch/instant-meilisearch) is a plugin connecting your Meilisearch instance with InstantSearch, giving you access to many (but not all) of the same front-end components as Algolia users. Here is [an up-to-date list of components compatible with Instant Meilisearch](https://github.com/meilisearch/instant-meilisearch/#-api-resources).

Reviewer 4: I would open this paragraph by saying something like "To use InstantSearch with Meilisearch, you need Instant Meilisearch" and so on.

Might be me, but I feel that right now the tools are presented as equivalent, when actually one depends on the other.

Replace `INDEX_NAME` with the name of the Algolia index you would like to migrate to Meilisearch.

### Fetch data from Algolia

To fetch all index data at once we can use Algolia's [`browseObjects`](https://www.algolia.com/doc/api-reference/api-methods/browse/) method.

```jsx
let records = [];
await algoliaIndex.browseObjects({
    batch: (hits) => {
      records=records.concat(hits)
    }
  });
```

On each batch of hits, the `batch` callback method is invoked, and the content is concatenated in the `records` array. This variable will be used later on in the upload process.

#### Nested objects

Meilisearch versions prior to v0.27.0 have limited support for nested fields, meaning that nested objects must be flattened manually. For this reason we recommend using [the latest version of Meilisearch](https://github.com/meilisearch/MeiliSearch/releases). If this is impossible, try the modified data retrieval process below.

Let's say your document structure looks like this:

```json
{
    "objectID": 1,
    "name": "John",
    "age": 20,
    "address": {
        "country": "Spain"
    }
}
```

> To export this data from Algolia and simultaneously flatten it, we will use the same `browseObjects` method as described above. However, before concatenation, we will modify the object to flatten `address` into `address.country`:

Reviewer 3: Actually I'm not sure why this is useful

```jsx
let records = [];
await algoliaIndex.browseObjects({
 batch: (recordsArr) => {
        records = recordsArr.reduce((acc, record) => {
          const { address, ...rest } = record;

>           acc.push({ "address.country": address.country, ...rest });

Reviewer 4: I think @bidoubiwa's solution is technically more complete and closer to real-life scenarios where data won't always conveniently formatted. But it is also more complex and risks distracting users who are more concerned about understanding the workflow than copypasting code.

>           acc.push({ "address.country": address.country, ...rest });

Reviewer 3: We discussed with @shivaylamba and decided to remove this part altogether. We discussed this subject a lot in the past and determined back then that we should probably not keep support for previous versions of Meilisearch (before the v1). We don't want users to keep older versions for their production because for example, older versions might have bugs that Meilisearch (the engine) will not fix.

It also adds a lot of complexity to this article for a very specific case. Nested is supported in v0.27.1, if the user wants I think that we should invite him to upgrade to this version.

>           acc.push({ "address.country": address.country, ...rest });

Author: I agree with this decision! Thanks @bidoubiwa and @shivaylamba 😊

### Upload data to Meilisearch

Next, use the Meilisearch [`addDocumentsInBatches`](/learn/core_concepts/documents.md#primary-field) method to upload all your records in batches of 1000 at a time.

```jsx
const BATCH_SIZE = 1000;
await meiliIndex.addDocumentsInBatches(records, BATCH_SIZE);
```

That's all! When you're ready to run the script, enter the below command:

```bash
node script.js
```

### Finished script

```jsx
const AlgoliaSearch = require("algoliasearch");
const { MeiliSearch } = require("meilisearch");
const BATCH_SIZE = 1000;

(async () => {
  const algoliaClient = AlgoliaSearch("APPLICATION_ID", "ADMIN_API_KEY");
  const algoliaIndex = algoliaClient.initIndex("INDEX_NAME");

  let records = [];
  await algoliaIndex.browseObjects({
    batch: (hits) => {
      records = records.concat(hits)
    }
  });

  const meiliClient = new MeiliSearch({
    host: "MEILI_HOST",
    apiKey: "MEILI_API_KEY",
  });
  const meiliIndex = meiliClient.index("MEILI_INDEX_NAME");

>   await meiliIndex.addDocumentsInBatches(records, BATCH_SIZE);

Author: How would you suggest changing the code sample?
